test/src/main/java/org/springframework/security/test/web/servlet/request/SecurityMockMvcRequestPostProcessors.java

```diff
@@ -1024,6 +1045,161 @@ public MockHttpServletRequest postProcessRequest(MockHttpServletRequest request)
+		 */
+		public OidcLoginRequestPostProcessor idToken(Consumer<OidcIdToken.Builder> idTokenBuilderConsumer) {
+			OidcIdToken.Builder builder = OidcIdToken.withTokenValue("id-token");
+			builder.subject("test-subject");
+			idTokenBuilderConsumer.accept(builder);
+			this.idToken = builder.build();
+			return this;
+		}
+
+		/**
+		 * Use the provided {@link OidcUserInfo} when constructing the authenticated user
+		 *
+		 * @param userInfoBuilderConsumer a {@link Consumer} of a {@link OidcUserInfo.Builder}
+		 * @return the {@link OidcLoginRequestPostProcessor} for further configuration
+		 */
+		public OidcLoginRequestPostProcessor userInfoToken(Consumer<OidcUserInfo.Builder> userInfoBuilderConsumer) {
+			OidcUserInfo.Builder builder = OidcUserInfo.builder();
+			userInfoBuilderConsumer.accept(builder);
+			this.userInfo = builder.build();
+			return this;
+		}
+
+		/**
+		 * Use the provided {@link OidcUser} as the authenticated user.
+		 *
+		 * Supplying an {@link OidcUser} will take precedence over {@link #idToken}, {@link #userInfo},
+		 * and list of {@link GrantedAuthority}s to use.
+		 *
+		 * @param oidcUser the {@link OidcUser} to use
+		 * @return the {@link OidcLoginRequestPostProcessor} for further configuration
+		 */
+		public OidcLoginRequestPostProcessor oidcUser(OidcUser oidcUser) {
+			this.oidcUser = oidcUser;
+			return this;
+		}
+
+		/**
+		 * Use the provided {@link ClientRegistration} as the client to authorize.
+		 *
+		 * The supplied {@link ClientRegistration} will be registered into an
+		 * {@link HttpSessionOAuth2AuthorizedClientRepository}. Tests relying on
+		 * {@link org.springframework.security.oauth2.client.annotation.RegisteredOAuth2AuthorizedClient}
+		 * annotations should register a {@link HttpSessionOAuth2AuthorizedClientRepository} bean
+		 * to the application context.
+		 *
+		 * The client registration must be a valid {@link ClientRegistration} from the
+		 * {@link org.springframework.security.oauth2.client.registration.ClientRegistrationRepository}
+		 * in the application context.
+		 *
+		 * @param clientRegistration the {@link ClientRegistration} to use
+		 * @return the {@link OidcLoginRequestPostProcessor} for further configuration
+		 */
+		public OidcLoginRequestPostProcessor clientRegistration(ClientRegistration clientRegistration) {
+			this.clientRegistration = clientRegistration;
+			return this;
+		}
+
+		@Override
+		public MockHttpServletRequest postProcessRequest(MockHttpServletRequest request) {
+			OidcUser oidcUser = getOidcUser();
+			OAuth2AuthenticationToken token = new OAuth2AuthenticationToken(oidcUser, oidcUser.getAuthorities(), this.clientRegistration.getRegistrationId());
+			OAuth2AuthorizedClient client = new OAuth2AuthorizedClient(this.clientRegistration, token.getName(), this.accessToken);
+			OAuth2AuthorizedClientRepository authorizedClientRepository = new HttpSessionOAuth2AuthorizedClientRepository();
```

Comment:
@jgrandja I looked into simply pulling this from the application context. The issue I discovered is that Spring Boot registers an `AuthenticatedPrincipalOAuth2AuthorizedClientRepository`, which will store this `client` in an `InMemoryOAuth2AuthorizedClientService`. Since that is likely shared between tests, it could cause a problem in concurrent tests.

I've chosen to stick with `HttpSessionOAuth2AuthorizedClientRepository` and documentation that indicates that tests should use that repository so that the argument resolver picks up the client created here. 

With this approach, `@RegisteredOAuth2AuthorizedClient` will work with still pretty minimal code in the test, e.g.:

```java
@Test 
public void test() {
    this.mvc.perform(get("/")
        .with(oidcLogin().clientRegistration(clientRegistration));
}

@Configuration
static class ClientConfig {
    @Bean 
    ClientRegistrationRepository clients() {
        return new InMemoryClientRegistrationRepository(clientRegistration);
    }

    @Bean 
    OAuth2AuthorizedClientRepository authorizedClients() {
        return new HttpSessionOAuth2AuthorizedClientRepository();
    }
}
```

This could be potentially reduced down the road if there's a simple, thread-safe way to reconfigure or override the `OAuth2AuthorizedClientArgumentResolver` within the confines of the test.

/cc @rwinch for additional thoughts.

Reply:
@jzheaux I like the way the test reads. It's simple, concise and consistent with `HttpSecurity.oauth2Login()` Servlet DSL. For the initial feature release, I think it's fine to require the use `HttpSessionOAuth2AuthorizedClientRepository` in tests. We can figure out ways to improve this later.